autobahn/websocket/types.py

```diff
@@ -127,6 +127,109 @@ def __str__(self):
         return json.dumps(self.__json__())
 
 
+@public
+class ConnectingRequest(object):
+    """
+    Thin-wrapper for WebSocket connection request information provided in
+    :meth:`autobahn.websocket.protocol.WebSocketClientProtocol.onConnecting`
+    after a client has connected, but before the handshake has
+    proceeded.
+
+    `host`, `port`, and `resource` are all required, everything else
+    is optional. Note that these are values that will be seen by the
+    client and should represent the public-facing host, port and
+    resource to which the client is connecting (not necessarily the
+    action host/port being used).
+    """
+
+    __slots__ = (
+        'host',
+        'port',
+        'resource',
+        'headers',
+        'useragent',
+        'origin',
+        'protocols',
+    )
+
+    def __init__(self, host=None, port=None, resource=None, headers=None, useragent=None, origin=None, protocols=None):
+        """
+        :param headers: HTTP headers to send in the opening handshake
+        :type headers: dict
+
+        """
+        # required
+        self.host = host if host is not None else "localhost"
+        self.port = port if port is not None else 80
+        self.resource = resource if resource is not None else "/"
+        # optional
+        self.headers = headers if headers else dict()
```

Comment:
Kind of an anti-pattern.  If `headers` is an empty dict then this sets `self.headers` to a _different_ empty dict.  It doesn't really matter for simple, straightforward code but it creates quite a surprise for nasty code that mutates that dict later on.

I'd suggest preserving identity:

```
self.headers = headers if headers is not None else dict()
```

And likewise for `protocols` below.

Reply:
ah, right! fwiw, the classes in "types" were envisioned as pure value holding classes - but the code as it stands is neither nor. if we'd go with "pure value types", I guess we should (deep)copy everything coming in via ctors. lets see  @meejah 's opinion ..

Reply:
We can definitely do what @exarkun suggests (which I agree is a good idea). At some point I had some "XXX" comments about this; we can "slow everything down" by deep-copying or make it "up to the caller" to make sure they're not mutating (or, only mutating in a way that makes sense?). We should definitely do the `is not None` thing ... I'm not so sure about deep-copying everything.

Reply:
ok, so no deep copy, and preserve identity for dict: +1

while discussing this, why is `dict()` better than `{}`?

Reply:
I prefer `dict()` to avoid confusion with `set()`s. A dict with one item: `{"foo": "bar"}` a set with two items: `{"foo", "bar"}`...you can't make an empty set with `{` and `}` AFAIK, but ....